runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/ArtifactStagingService.java

```diff
@@ -0,0 +1,606 @@
+            path = RunnerApi.ArtifactUrlPayload.parseFrom(artifact.getTypePayload()).getUrl();
+          } else {
+            path = "artifact";
+          }
+        } catch (InvalidProtocolBufferException exn) {
+          throw new RuntimeException(exn);
+        }
+        // Limit to the last contiguous alpha-numeric sequence. In particular, this will exclude
+        // all path separators.
+        List<String> components = Splitter.onPattern("[^A-Za-z-_.]]").splitToList(path);
+        String base = components.get(components.size() - 1);
+        return clip(String.format("%d-%s-%s", index, clip(environment, 25), base), 100);
+      }
+
+      private String clip(String s, int maxLength) {
+        return s.length() < maxLength ? s : s.substring(0, maxLength);
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        stagingExecutor.shutdownNow();
+        LOG.error("Error staging artifacts", throwable);
+        state = State.ERROR;
+      }
+
+      @Override
+      public void onCompleted() {
+        assert state == State.DONE;
+      }
+    };
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Nothing to close.
+  }
+
+  /**
+   * Lazily stages artifacts by letting an ArtifactStagingService resolve and request artifacts.
+   *
+   * @param retrievalService an ArtifactRetrievalService used to resolve and retrieve artifacts
+   * @param stagingService an ArtifactStagingService stub which will request artifacts
+   * @param stagingToken the staging token of the job whose artifacts will be retrieved
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  public static void offer(
+      ArtifactRetrievalService retrievalService,
+      ArtifactStagingServiceGrpc.ArtifactStagingServiceStub stagingService,
+      String stagingToken)
+      throws InterruptedException, IOException {
+    StagingRequestObserver requestObserver = new StagingRequestObserver(retrievalService);
+    requestObserver.responseObserver =
+        stagingService.reverseArtifactRetrievalService(requestObserver);
+    requestObserver.responseObserver.onNext(
+        ArtifactApi.ArtifactResponseWrapper.newBuilder().setStagingToken(stagingToken).build());
+    requestObserver.waitUntilDone();
+    if (requestObserver.error != null) {
+      if (requestObserver.error instanceof IOException) {
+        throw (IOException) requestObserver.error;
+      } else {
+        throw new IOException(requestObserver.error);
+      }
+    }
+  }
+
+  /** Actually implements the reverse retrieval protocol. */
+  private static class StagingRequestObserver
+      implements StreamObserver<ArtifactApi.ArtifactRequestWrapper> {
+
+    private ArtifactRetrievalService retrievalService;
+
+    public StagingRequestObserver(ArtifactRetrievalService retrievalService) {
+      this.retrievalService = retrievalService;
+    }
+
+    CountDownLatch latch = new CountDownLatch(1);
```

Comment:
You could use a CompletableFuture<Void> instead of `latch` and `error`. In the onCompleted/onError methods you complete the future. `waitUntilDone` would then just be `future.get`